From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: etil2jz <81570777+etil2jz@users.noreply.github.com>
Date: Tue, 2 Aug 2022 14:48:12 +0200
Subject: [PATCH] Mirai: Configurable chat message signatures

Fixed & Updated by Dreeam-qwq
Original license: GPL-3.0-only
Original project: https://github.com/Dreeam-qwq/Mirai

Original license: GPL-3.0-only
Original project: https://github.com/etil2jz/Mirai

diff --git a/net/minecraft/network/protocol/game/ClientboundLoginPacket.java b/net/minecraft/network/protocol/game/ClientboundLoginPacket.java
--- a/net/minecraft/network/protocol/game/ClientboundLoginPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundLoginPacket.java
@@ -49,7 +49,7 @@ public record ClientboundLoginPacket(
         ByteBufCodecs.BOOL,
         ClientboundLoginPacket::onlineMode,
         ByteBufCodecs.BOOL,
-        ClientboundLoginPacket::enforcesSecureChat,
+        packet -> !org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled || packet.enforcesSecureChat(), // Leaf - Mirai - Configurable chat message signatures
         ClientboundLoginPacket::new
     );
 
diff --git a/net/minecraft/network/protocol/game/ServerboundChatPacket.java b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -20,7 +20,7 @@ public record ServerboundChatPacket(String message, Instant timeStamp, long salt
         ByteBufCodecs.LONG,
         ServerboundChatPacket::salt,
         MessageSignature.STREAM_CODEC.apply(ByteBufCodecs::optional),
-        ServerboundChatPacket::signature,
+        packet -> org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled ? packet.signature() : Optional.empty(), // Leaf - Mirai - Configurable chat message signatures
         LastSeenMessages.Update.STREAM_CODEC,
         ServerboundChatPacket::lastSeenMessages,
         ServerboundChatPacket::new
diff --git a/net/minecraft/network/protocol/status/ServerStatus.java b/net/minecraft/network/protocol/status/ServerStatus.java
--- a/net/minecraft/network/protocol/status/ServerStatus.java
+++ b/net/minecraft/network/protocol/status/ServerStatus.java
@@ -20,9 +20,13 @@ public record ServerStatus(
     Optional<ServerStatus.Version> version,
     Optional<ServerStatus.Favicon> favicon,
     boolean enforcesSecureChat
+    , boolean preventsChatReports // Leaf - Mirai - Configurable chat message signatures
 ) {
     public static final Codec<ServerStatus> CODEC = RecordCodecBuilder.create(
-        i -> i.group(
+        // Leaf start - Mirai - Configurable chat message signatures
+        i ->
+            org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled
+            ? i.group(
                 ComponentSerialization.CODEC.lenientOptionalFieldOf("description", CommonComponents.EMPTY).forGetter(ServerStatus::description),
                 ServerStatus.Players.CODEC.lenientOptionalFieldOf("players").forGetter(ServerStatus::players),
                 ServerStatus.Version.CODEC.lenientOptionalFieldOf("version").forGetter(ServerStatus::version),
@@ -30,8 +34,30 @@ public record ServerStatus(
                 Codec.BOOL.lenientOptionalFieldOf("enforcesSecureChat", false).forGetter(ServerStatus::enforcesSecureChat)
             )
             .apply(i, ServerStatus::new)
+            : i.group(
+                ComponentSerialization.CODEC.lenientOptionalFieldOf("description", CommonComponents.EMPTY).forGetter(ServerStatus::description),
+                ServerStatus.Players.CODEC.lenientOptionalFieldOf("players").forGetter(ServerStatus::players),
+                ServerStatus.Version.CODEC.lenientOptionalFieldOf("version").forGetter(ServerStatus::version),
+                ServerStatus.Favicon.CODEC.lenientOptionalFieldOf("favicon").forGetter(ServerStatus::favicon),
+                Codec.BOOL.lenientOptionalFieldOf("enforcesSecureChat", false).forGetter(x -> true),
+                Codec.BOOL.lenientOptionalFieldOf("preventsChatReports", false).forGetter(x -> true)
+                )
+            .apply(i, ServerStatus::new)
+            // Leaf end- Mirai - Configurable chat message signatures
     );
 
+    // Leaf start - Mirai - Configurable chat message signatures
+    public ServerStatus(
+        final Component description,
+        final Optional<ServerStatus.Players> players,
+        final Optional<ServerStatus.Version> version,
+        final Optional<ServerStatus.Favicon> favicon,
+        final boolean enforcesSecureChat
+    ) {
+        this(description, players, version, favicon, enforcesSecureChat, false);
+    }
+    // Leaf end- Mirai - Configurable chat message signatures
+
     public record Favicon(byte[] iconBytes) {
         private static final String PREFIX = "data:image/png;base64,";
         public static final Codec<ServerStatus.Favicon> CODEC = Codec.STRING.comapFlatMap(string -> {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -814,6 +814,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public boolean enforceSecureProfile() {
+        if (!org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled) return false; // Leaf - Mirai - Configurable chat message signatures
         DedicatedServerProperties properties = this.getProperties();
         // Paper start - Add setting for proxy online mode status
         return properties.enforceSecureProfile
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -2956,10 +2956,22 @@ public class ServerGamePacketListenerImpl
 
     public void sendPlayerChatMessage(final PlayerChatMessage message, final ChatType.Bound chatType) {
         // CraftBukkit start - SPIGOT-7262: if hidden we have to send as disguised message. Query whether we should send at all (but changing this may not be expected).
-        if (!this.getCraftPlayer().canSeePlayer(message.link().sender())) {
-            this.sendDisguisedChatMessage(message.decoratedContent(), chatType);
+        // Leaf start - Mirai - Configurable chat message signatures
+        final boolean canSeeSender = this.getCraftPlayer().canSeePlayer(message.link().sender());
+        if (!canSeeSender || !org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled) {
+            final @Nullable Component content = message.filterMask().isEmpty()
+                ? message.decoratedContent()
+                : message.filterMask().applyWithFormatting(message.signedContent());
+            if (content != null) {
+                if (canSeeSender) {
+                    this.send(new ClientboundSystemChatPacket(chatType.decorate(content), false));
+                } else {
+                    this.sendDisguisedChatMessage(content, chatType);
+                }
+            }
             return;
         }
+        // Leaf end - Mirai - Configurable chat message signatures
         // CraftBukkit end
         // Paper start - Ensure that client receives chat packets in the same order that we add into the message signature cache
         synchronized (this.messageSignatureCache) {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1239,7 +1239,7 @@ public abstract class PlayerList {
     public void broadcastChatMessage(final PlayerChatMessage message, final Predicate<ServerPlayer> isFiltered, final @Nullable ServerPlayer senderPlayer, final ChatType.Bound chatType, final @Nullable Function<net.kyori.adventure.audience.Audience, Component> unsignedFunction) {
         // Paper end
         boolean trusted = this.verifyChatTrusted(message);
-        this.server.logChatMessage((unsignedFunction == null ? message.decoratedContent() : unsignedFunction.apply(this.server.console)), chatType, trusted || !org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.chat.notSecureMarker ? null : "Not Secure"); // Paper // Gale - do not log Not Secure marker
+        this.server.logChatMessage((unsignedFunction == null ? message.decoratedContent() : unsignedFunction.apply(this.server.console)), chatType, trusted || !org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled || !org.galemc.gale.configuration.GaleGlobalConfiguration.get().logToConsole.chat.notSecureMarker ? null : "Not Secure"); // Paper // Gale - do not log Not Secure marker // Leaf - Mirai - Configurable chat message signatures
         OutgoingChatMessage tracked = OutgoingChatMessage.create(message);
         boolean wasFullyFiltered = false;
 
@@ -1264,6 +1264,7 @@ public abstract class PlayerList {
     }
 
     public boolean verifyChatTrusted(final PlayerChatMessage message) {
+        if (!org.dreeam.leaf.config.modules.network.ChatMessageSignature.enabled) return true; // Leaf - Mirai - Configurable chat message signatures
         return message.hasSignature() && !message.hasExpiredServer(Instant.now());
     }
 
